### DISCUSSION TITLE
Branching with self hosted Neon

### DISCUSSION THREAD

**Discussion by mingfang** [neondatabase/neon#3413] [General]
Is it possible to do branching with a self hosted Neon setup?

In self hosted, we only have access to the pageserver api 
https://github.com/neondatabase/neon/blob/main/pageserver/src/http/openapi_spec.yml
and I don't see any api for branching.

**Comment by SomeoneToIgnore** [comment]
Branching is possible with `POST /v1/tenant/{tenant_id}/timeline/` query:

https://github.com/neondatabase/neon/blob/aabca55d7e7e020a30f6afd286c850af88063f28/pageserver/src/http/openapi_spec.yml#L503-L508

There are two ways to create a timeline:
* from scratch, then no `ancestor_timeline_id` and `ancestor_start_lsn` are specified
* from another existing timeline (aka branching), then at least `ancestor_timeline_id` should be specified for that. If something else than "latest current state of the ancestor timeline" is needed to branch off, `ancestor_start_lsn` could be specified.

**Reply by mingfang** [reply]
@SomeoneToIgnore 
This is great. Thanks.
